packages/components/src/custom-select-control-v2/types.ts

```diff
@@ -22,7 +27,13 @@ export type CustomSelectProps = {
 	 * An optional default value for the control. If left `undefined`, the first
 	 * non-disabled item will be used.
 	 */
-	defaultValue?: string | string[];
+	defaultValue: string | string[];
```

Comment:
When adding `unmountOnHide`, the initial selected value was an empty string because it no longer had access to the items through the store (see details here: https://github.com/ariakit/ariakit/issues/3374#issuecomment-1905072489)

I'm thinking it might be cleanest to require a `defaultValue` (for the new version, the legacy would use the hardcoded default to behave the same), as it's just a string for the consumer to implement. 

Otherwise, we'd have to map over the children to find a default value. The problem here is that the children could be a fragment or a div or anything that the consumer implements. So it seems like it could get a bit messy to extract a value. 

It's also likely I'm overcomplicating it and there is an easy way to get the value. I'm happy to hear if there is a better solution. 🙂

Reply:
Is `defaultValue` still required if the `value` prop is provided?

Reply:
As Diego suggests, `defaultValue` shouldn't be required when `value` is provided — it would actually cause more confusion, since `defaultValue` is used to provide an initial value to the component when used as uncontrolled, while `value` is used to set the component's value in a controlled mode.

> Otherwise, we'd have to map over the children to find a default value. The problem here is that the children could be a fragment or a div or anything that the consumer implements

If we expect children to be instances of `CustomSelectControlItem`, then we could read the `value` prop, which is required on those components, and assume as a default value the first defined `value` that we find when mapping over the children.

Would that work?

Reply:
Is there a way we can access the `value` prop without mapping through the children? 

That is the only way I am aware of. Other than doing something that might be considered strange and passing it from the child to parent through context. 

If we map though the children to find the `value`, then we're back to the same problem I shared earlier where we might not know how many layers deep the value is. For example, if they are wrapped in a `div`: 

<img width="616" alt="Screenshot 2024-01-25 at 3 45 12 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/3f8041a5-a8a7-420c-8566-8c7cdc61c23b">

Unless we have a recursive function but that's where I thought it was getting messy or that I was overcomplicating it.

Reply:
Yeah, it's tricky. If the `CustomSelectControlItem` component gets abstracted into another custom component, you won't have access to it, even with recursion. It seems odd to impose such artificial restrictions on JSX, where users expect certain features to work seamlessly. It's like trying to use JSX as a DSL for an array of objects, when simply using the array of objects might be the answer.

I'd suggest that either `defaultValue` or `value` should be required. There are ways to express this in TypeScript, but a runtime warning might be necessary because the TypeScript error may not be very clear.

Reply:
Thanks for the suggestion! 

There is another conflict due to `unmountOnHide` — since we don't have access to the values, entering characters via keydown no longer works unless the popover is open/mounted. This behavior makes sense with how the ariakit component works, but it conflicts with the legacy behavior. And I think it's a nice feature to have in the new version. 

In our legacy version, we have a fallback added for this scenario: 

https://github.com/WordPress/gutenberg/blob/fa802d10e65c7d0951cc0f2357c367b0ab19861c/packages/components/src/custom-select-control/index.js#L27-L64

However, with the limitation of using compound components, this would be complicated to implement.

In addition to this, there are other related test failures (that I still need to look into, but have confirmed it's a result of the `unmountOnHide` prop). 

So it's getting a bit messy, but perhaps I'm overcomplicating this again. Curious to hear anyone's thoughts and/or ideas on this!

Reply:
Update:

@mirka and I took a deeper look and observed a few things:

- the native `select` element does support typeahead, but it opens the dropdown popover when pressing the arrow keys
- it looks like `ariakit` doesn't support typeahead natively when `unmountOnHide` is `true` (cc @diegohaz , is this something that ariakit should fix ?)
- however, Lena suggested a little hack which seems to work: set `unmountOnHide` to `false` when focusing the button that is triggering the dropdown (see [example](https://stackblitz.com/edit/utdlzl?file=select%2Fstyle.css,select%2Findex.tsx&theme=dark))

With respect to the initial value, it should be determined by the consumer by passing either the `value` prop (controlled mode) or the `defaultValue` prop (uncontrolled mode)

Reply:
>however, Lena suggested a little hack which seems to work: set unmountOnHide to false when focusing the button that is triggering the dropdown 

Great idea!

Reply:
> but the test for this behavior is still failing:

From personal experience, `user-event` may cause some incompatibilities with ariakit-based components, especially when we make some async changes like in the suggestion above. Maybe switching to `@ariakit/test` fixes it?

> Could this be a potential solution?

It could work, as an additional hack. I've updated the example and it [seems to work](https://stackblitz.com/edit/utdlzl?file=select%2Fstyle.css,select%2Findex.tsx&theme=dark) as expected. It may have some performance ripercussion, though — so we could go with this implementation at first (which allows us to keep `defaultValue` as optional), and only switch to requiring `defaultValue` as a fallback? What do folks think?

Reply:
>Maybe switching to @ariakit/test fixes it?

It does! I just thought we might want to keep the tests as close to the original `CustomSelectControl` as possible, so I planned to move to `ariakit/test` after. But since the tests are the same aside from the interaction method, maybe it will be okay to do that now.

Reply:
> If we mount and then unmount it on the initial render

I feel like the initial render is exactly the most important time to reap the performance benefits of `unmountOnHide`, so I'm not sure it makes sense to mount on initial render but hide it after.

So going back to basics, if we don't already have proof that just keeping `unmountOnHide={ false }` _all the time_ has a perceivable performance hit to the overall Gutenberg app, I'd even be fine with not messing with that prop at all for the first iteration. I have to imagine that there are much fewer CustomSelectControls in any given view than a generic Tooltip, for example.

Reply:
So, we remove `unmountOnHide` and all of the related workarounds? 🙈 If I understood correctly, then that is done through: https://github.com/WordPress/gutenberg/pull/57902/commits/6baa6426c7e2cc12052a2779807ac0d1767f716b

Just to note: the tests will pass using `@testing-library/user-event` or with `@ariakit/test`. I believe the former is faster, but I think we will likely need to move to `@ariakit/test` when integrating the new tests (https://github.com/WordPress/gutenberg/pull/58583) — based on seeing flaky test results when merging the new and legacy tests in another version of the legacy adapter: https://github.com/WordPress/gutenberg/pull/57000

We could revert to RTL, merge this PR, and then merge the tests/change to `@ariakit/test` if needed in a follow-up. Or we could keep it with `@ariakit/test`, and I can merge the new tests into this. 

Any preferences?